service/implementation/email_sender/user_subscription_manager.py

```diff
@@ -0,0 +1,99 @@
+
+from sqlalchemy import event
+
+from api.container.container import Container
+from database.postgres.dal.user_subscription import UserSubscriptionDAL
+
+
+class UserSubscriptionManagerMeta(type):
+    _instances = {}
+
+    def __call__(cls, *args, **kwargs):
+        if cls not in cls._instances:
+            _instance = super().__call__(*args, **kwargs)
+            cls._instances[cls] = _instance
+        return cls._instances[cls]
+
+class UserSubscriptionManager(metaclass=UserSubscriptionManagerMeta):
+    __SCOPES = ["https://www.googleapis.com/auth/gmail.send"]
+
+    @inject
+    def __init__(self, user_subscription_dal: UserSubscriptionDAL = Provide[Container.user_subscription_dal]):
+        self.__user_subscription_dal = user_subscription_dal
+
+        current_dir = Path(__file__).resolve().parent if "__file__" in locals() else Path.cwd()
+        envars = current_dir / ".env"
+        load_dotenv(envars)
+
+        self.__sender_email = os.getenv("EMAIL")
+        self.__sender_password = os.getenv("PASSWORD")
+
+        event.listen(TempSubscribersData, "after_insert", self.__on_subscribers_inserted)
+
+    @inject
+    def try_add_subscribers_to_temp_table(self, team_index):
+        self.__user_subscription_dal.try_add_subscribers_data(team_index)
+
+    def __on_subscribers_inserted(self, mapper, connection, target):
+        self.__try_send_email_to_users()
+
+    def __try_send_email_to_users(self):
+        subscribed_users = self.__user_subscription_dal.get_subscribed_users_data_and_delete_rows()
+
+        email_subject = "News notification"
+        email_body = "The news with the team that you subscribed in have just been released"
+
+        for user in subscribed_users:
+            self.__send_email_to_user(user.subscriber_emails, email_subject, email_body)
+
+    def __send_email_to_user(self, user_email: str, subject, body):
+        creds = self.__authenticate_gmail()
+        service = build("gmail", "v1", credentials=creds)
+
+        message = MIMEText(body)
+        message["to"] = user_email
+        message["subject"] = subject
+        raw_message = base64.urlsafe_b64encode(message.as_bytes()).decode()
+
+        send_message = service.users().messages().send(
+            userId="me",
+            body={"raw": raw_message}
+        ).execute()
+
+        print(f"Email sent successfully! Message ID: {send_message['id']}")
+
+    def __authenticate_gmail(self):
+        creds = None
+
+        if os.path.exists("token.json"):
+            creds = Credentials.from_authorized_user_file("token.json", self.__SCOPES)
+
+        if not creds or not creds.valid:
+            if creds and creds.expired and creds.refresh_token:
+                creds.refresh(Request())
+            else:
+                credentials_path = r"C:\Users\LEGION\Documents\Softserve\SoftServeProject\credentials.json"
+                flow = InstalledAppFlow.from_client_secrets_file(credentials_path, self.__SCOPES)
```

Comment:
We absolutely MUST not work with Google Auth on our server. 


Here's some info from our GPT friend:

No, InstalledAppFlow is for desktop or local apps, not Web APIs.

For a Web API, you should use Authorization Code Flow with google-auth and oauthlib. You'd:

Redirect users to Google's OAuth consent page.
Handle the callback with an authorization code.
Exchange the code for access tokens on your backend.
Use google-auth-oauthlib.flow.Flow.from_client_secrets_file(...) instead, with Flow.authorization_url() and Flow.fetch_token().